provide a VCS stash segment

dffischer opened the pull request:
Some version control systems have an area where changes can be stored as temporary work in progress instead of publishing them to the history, such as Git's "stash" or Mercurial's "shelve". Their count neatly displayed in powerlines similar to the branch name could remind a developer that there are still open ends left in his repository.

This pull requests sets up a segment doing so and provides an implementation for Git repositories. As it is my first take on the powerline source code, comments are much appreciated.

ZyX-I commented on powerline/lib/vcs/git.py:
```diff
@@ -102,6 +102,13 @@ class Repository(GitRepository):
+		def stash(self):
+			try:
+				return sum(1 for _ in git.Repository(git_directory(self.directory))
+					.lookup_reference('refs/stash').log())
```
`.attr` should not be on the next line, also there are too many calls under `try`. You should use

    try:
        stashref = git.Repository(git_directory(self.directory)).lookup_reference('refs/stash')
    except KeyError:
        return None
    return sum(1 for _ in stashref.log())

ZyX-I commented on powerline/config_files/themes/shell/default.json:
```diff
@@ -31,6 +31,10 @@
 			},
 			{
+				"function": "powerline.segments.common.vcs.stash",
+				"priority": 30
```
This priority is higher then the one of the branch segment, not lower.

ZyX-I commented on powerline/config_files/colorschemes/default.json:
```diff
@@ -43,6 +43,8 @@
 		"branch_dirty":              { "fg": "brightyellow", "bg": "gray2", "attrs": [] },
 		"branch_clean":              { "fg": "gray9", "bg": "gray2", "attrs": [] },
 		"branch:divider":            { "fg": "gray7", "bg": "gray2", "attrs": [] },
+		"stash":                     { "fg": "brightyellow", "bg": "gray2", "attrs": [] },
```
If `stash` is a copy of `branch_dirty` everywhere it is much better to use `"stash": "branch_dirty"` where it is a copy. Same for `stash:divider`, this also avoids changing anything in Vim colorschemes.

ZyX-I commented on powerline/config_files/colorschemes/vim/__main__.json:
```diff
@@ -7,6 +7,7 @@
 		"file_encoding":      "file_format",
 		"file_type":          "file_format",
 		"branch":             "information:additional",
+		"stash":              "information:additional",
```
Either this or `"stash"` entries in Vim colorschemes must not be present. It is absolutely useless to say “stash should be highlighted just as information:additional” just to then state “stash should be highlighted with fg, bg, attrs”.

ZyX-I replied:
Though with top-level aliasing this entry will also be not needed.

dffischer replied:
I did understand the string assignments in color schemes as categories instead of aliases. Looks like they provide exactly what I intended.

The *default* color scheme defines `branch_dirty` and `branch:divider` in the global scheme, so I aliased it there. In *solarized*, `stash:divider` is only defined in the definition for Vim, so declared the alias there, too.

ZyX-I commented on powerline/segments/common/vcs.py:
```diff
@@ -56,3 +56,34 @@ def __call__(self, pl, segment_info, create_watcher, status_colors=False, ignore
+stash = with_docstring(StashSegment(),
+'''Return the number of current VCS stash entries, if any.
+
+Highlight groups used: ``stash``.
```
Missing note about divider highlight group.

dffischer replied:
Here, I must object. The divider highlight group is set to `None` at the top of the `StashSegment` class. It and then only utilized in the Vim variant of the segment, where already properly documented, analogous to the `BranchSegment`.

ZyX-I replied:
Ah, OK.

ZyX-I commented on powerline/segments/vim/__init__.py:
```diff
@@ -512,6 +512,27 @@ def get_directory(segment_info):
+branch = with_docstring(VimStashSegment(),
+'''Return the number of stashes in the current working branch.
+
+Highlight group used: ``stash``.
```
Must use `groups`, not `group`. This information is for the linter, not only for the user.

ZyX-I commented on powerline/segments/vim/__init__.py:
```diff
@@ -512,6 +512,27 @@ def get_directory(segment_info):
+		return buffer_name(segment_info)
+
+
+branch = with_docstring(VimStashSegment(),
```
Incorrect variable name, this fails tests.

dffischer commented:
Thanks for the helpful pointers. I incorporated all of them but one and updated the branch. In some checks of the color schemes in my running system the linter seems happy.

ZyX-I commented on powerline/config_files/colorschemes/default.json:
```diff
@@ -43,6 +43,8 @@
 		"branch_clean":              { "fg": "gray9", "bg": "gray2", "attrs": [] },
 		"branch:divider":            { "fg": "gray7", "bg": "gray2", "attrs": [] },
+    "stash":                     "branch_dirty",
+    "stash:divider":             "branch:divider",
```
You have spaces here.

ZyX-I commented on powerline/config_files/colorschemes/solarized.json:
```diff
@@ -15,6 +15,7 @@
 		"branch":                  { "fg": "solarized:base1", "bg": "solarized:base02", "attrs": [] },
 		"branch_dirty":            { "fg": "solarized:yellow", "bg": "solarized:base02", "attrs": [] },
 		"branch_clean":            { "fg": "solarized:base1", "bg": "solarized:base02", "attrs": [] },
+    "stash":                   "branch_dirty",
```
And here.

ZyX-I commented on powerline/config_files/colorschemes/vim/solarized.json:
```diff
@@ -12,6 +12,7 @@
 		"readonly_indicator":       { "fg": "solarized:red", "bg": "solarized:base01", "attrs": [] },
 		"branch_dirty":             { "fg": "solarized:yellow", "bg": "solarized:base01", "attrs": [] },
 		"branch:divider":           { "fg": "solarized:base1", "bg": "solarized:base01", "attrs": [] },
+    "stash:divider":            "branch:divider",
```
And here.

ZyX-I commented on powerline/config_files/colorschemes/vim/solarizedlight.json:
```diff
@@ -12,6 +12,8 @@
 		"branch_dirty":             { "fg": "solarized:yellow", "bg": "solarized:base2", "attrs": [] },
 		"branch:divider":           { "fg": "solarized:base1", "bg": "solarized:base2", "attrs": [] },
+    "stash":                    "branch_dirty",
+    "stash:divider":            "branch:divider",
```
And here.

ZyX-I commented:
Still, tests: see tests/test_lib.py / TestVCS class and tests/test_segments.py / TestVcs, TestVim. Also I am not sure that Vim segment should exist at all, without file-watcher-based caching stash segment should be slow for Vim statusline.

dffischer commented:
Tests were added, spacing corrected.
I am not overly attached to the Vim variant. The branch segment had one, too, so I just added it for completeness.

ZyX-I commented on tests/test_lib.py:
```diff
@@ -566,6 +566,31 @@ def test_git(self):
+			return call(['git', 'stash', 'drop'], cwd=GIT_REPO, stdout=PIPE)
+
+		def stash_list():
+			return call(['git', 'stash', 'list'], cwd=GIT_REPO, stdout=PIPE)
```
AFAIR linter requires new line here.

ZyX-I commented on tests/test_segments.py:
```diff
@@ -738,6 +738,28 @@ def test_branch(self):
+		stash = partial(self.module.stash, pl=pl, create_watcher=create_watcher, segment_info={'getcwd': os.getcwd})
+
+		def forge_stash(n):
+		    return replace_attr(self.module, 'guess', get_dummy_guess(stash=lambda: n, directory='/tmp/tests'))
```
And here (around functions).

ZyX-I commented on tests/test_segments.py:
```diff
@@ -1390,6 +1412,29 @@ def test_branch(self):
+			stash = partial(self.vim.stash, pl=pl, create_watcher=create_watcher, segment_info=segment_info)
+
+			def forge_stash(n):
+				return replace_attr(self.vcs, 'guess', get_dummy_guess(stash=lambda: n))
```
New line.

ZyX-I commented on tests/test_lib.py:
```diff
@@ -566,6 +566,31 @@ def test_git(self):
+			stash_drop()
+			self.assertEqual(repo.stash(), 1)
+			stash_drop()
+			self.assertEqual(repo.stash(), None)
```
It returns 0 and not None and this fails all tests.

ZyX-I commented on tests/test_segments.py:
```diff
@@ -738,6 +738,28 @@ def test_branch(self):
+
+		def forge_stash(n):
+		    return replace_attr(self.module, 'guess', get_dummy_guess(stash=lambda: n, directory='/tmp/tests'))
+		with forge_stash(None):
```
And `None` is never returned by `.stash()`.

ZyX-I commented on tests/test_segments.py:
```diff
@@ -1390,6 +1412,29 @@ def test_branch(self):
+
+			def forge_stash(n):
+				return replace_attr(self.vcs, 'guess', get_dummy_guess(stash=lambda: n))
+			with forge_stash(None):
```
0

dffischer commented:
When called like mentioned in the [contribution guidelines](https://github.com/powerline/powerline/blob/develop/CONTRIBUTING.rst#programming-style), the linter only suggests newlines *before* functions. I added them also after them.

The pygit2 variant of the stash function returned `None` instead of `0`. This is now adapted to the variant using the Git command line program, which is used without pygit2 installed.

dffischer commented:
Thank you for the quick feedback and swift merge.
